src/lib/utilities/sokoban.ts

```diff
@@ -0,0 +1,512 @@
+			if (coordinatedGameComponents[i].component === EmojiGameComponent.NewLine) {
+				const gap = maxColumnIndice - coordinatedGameComponents[i].x;
+				coordinatedGameComponents.splice(i, 0, ...Array(gap).fill({ x: 0, y: 0, component: EmojiGameComponent.Null }));
+				i += gap;
+			}
+		}
+
+		this.gameComponents = coordinatedGameComponents.map((c) => c.component);
+	}
+
+	/**
+	 * Gets a component at the given position.
+	 *
+	 * @param x x-axis position.
+	 * @param y y-axis position.
+	 * @returns the component at the given position.
+	 */
+	public get(x: number, y: number): GameComponent {
+		return { component: this.gameComponents[y * (this.columnIndices + 1) + x], x, y };
+	}
+
+	/**
+	 * Sets a component at the given position.
+	 *
+	 * @param x x-axis position.
+	 * @param y y-axis position.
+	 * @param component the component to set the position to.
+	 */
+	public set(x: number, y: number, component: EmojiGameComponent) {
+		this.gameComponents[y * (this.columnIndices + 1) + x] = component;
+	}
+
+	/**
+	 * Gets the player component.
+	 *
+	 * @returns the player component.
+	 */
+	public getPlayer(): GameComponent {
+		let x = 0;
+		let y = 0;
+		const component = this.gameComponents.find((component) => {
+			if (component === EmojiGameComponent.NewLine) {
+				y++;
+				x = 0;
+				return false;
+			}
+			x++;
+			if ([EmojiGameComponent.Player, EmojiGameComponent.PlayerTarget].includes(component)) {
+				return true;
+			}
+			return false;
+		})!;
+		return { component, x: Math.max(0, x - 1), y };
+	}
+
+	/**
+	 * Peeks the next component for a viable move in the given direction.
+	 *
+	 * @param precedingComponent the component to peek around.
+	 * @param direction the direction to peek in.
+	 * @param boxRecurse whether the function is being recursed by a box.
+	 * @returns the next component if it's a viable move.
+	 */
+	public peekNextComponent(precedingComponent: GameComponent, direction: Direction, boxRecurse = false): Option<GameComponent> {
+		const nextX = precedingComponent.x + (direction === Direction.Left ? -1 : direction === Direction.Right ? 1 : 0);
+		const nextY = precedingComponent.y + (direction === Direction.Up ? -1 : direction === Direction.Down ? 1 : 0);
+
+		const nextComponent = this.get(nextX, nextY);
+		if (!nextComponent.component) return Option.none;
+		if ([EmojiGameComponent.Wall, EmojiGameComponent.Empty, EmojiGameComponent.Null].includes(nextComponent.component)) return Option.none;
+		if ([EmojiGameComponent.Box, EmojiGameComponent.BoxTarget].includes(nextComponent.component)) {
+			if (boxRecurse) return Option.none;
+			const nextNextComponent = this.peekNextComponent(nextComponent, direction, true);
+			if (nextNextComponent.isNone()) return Option.none;
+		}
+
+		return Option.some(nextComponent);
+	}
+
+	/**
+	 * Checks for nonviable moves.
+	 *
+	 * @returns the directions the player can't move in.
+	 */
+	public checkNonviableMoves(): Direction[] {
```

Comment:
This is an interesting choice, I usually see `getPossibleMoves`, but I rarely see negatives in code.

Reply:
this was intentional for the disabling of buttons in `buildGameControls()` via the second param `disabledDirections`, but now that you mentioned this, I realize I can just flip the boolean on the button disabling. 😅